tools/spark-block-cleaner/src/main/scala/org/apache/kyuubi/tools/KubernetesSparkBlockCleaner.scala

```diff
@@ -0,0 +1,205 @@
+          // delete empty sub directory
+          checkAndDeleteFile(subDir, time, true)
+          subDirReleased.sum
+        }
+        // delete empty blockManager directory
+        checkAndDeleteFile(blockManagerDir, time, true)
+        info(s"finished clean blockManager dir ${blockManagerDir.getCanonicalPath}, " +
+          s"released space: ${released.sum / 1024 / 1024} MB.")
+      }
+
+    // clean spark cache file
+    dir.listFiles.filter(_.isDirectory).filter(_.getName.startsWith("spark"))
+      .foreach { cacheDir =>
+        info(s"start check cache dir ${cacheDir.getCanonicalPath}")
+        val released = cacheDir.listFiles.map(file => checkAndDeleteFile(file, time))
+        // delete empty spark cache file
+        checkAndDeleteFile(cacheDir, time, true)
+        info(s"finished clean cache dir ${cacheDir.getCanonicalPath}, " +
+          s"released space: ${released.sum / 1024 / 1024} MB.")
+      }
+  }
+
+  private def checkAndDeleteFile(file: File, time: Long, isDir: Boolean = false): Long = {
+    debug(s"check file ${file.getName}")
+    val shouldDeleteFile = if (isDir) {
+      file.listFiles.isEmpty && (System.currentTimeMillis() - file.lastModified() > time)
+    } else {
+      System.currentTimeMillis() - file.lastModified() > time
+    }
+    val length = if (isDir) 0 else file.length()
+    if (shouldDeleteFile) {
+      if (file.delete()) {
+        debug(s"delete file ${file.getAbsolutePath} success")
+        return length
+      } else {
+        warn(s"delete file ${file.getAbsolutePath} fail")
+      }
+    }
+    0L
+  }
+
+  import scala.sys.process._
+
+  private def needToDeepClean(dir: String): Boolean = {
+    val used = (s"df $dir" #| s"grep $dir").!!
+      .split(" ").filter(_.endsWith("%")) {
+      0
+    }.replace("%", "")
+    info(s"$dir now used $used% space")
+
+    used.toInt > (100 - freeSpaceThreshold)
+  }
+
+  private def doCleanJob(dir: String): Unit = {
+    val startTime = System.currentTimeMillis()
+    val path = Paths.get(dir)
+    info(s"start clean job for $dir")
+    doClean(path.toFile, fileExpiredTime)
+    // re check if the disk has enough space
+    if (needToDeepClean(dir)) {
+      info(s"start deep clean job for $dir")
+      doClean(path.toFile, deepCleanFileExpiredTime)
+      if (needToDeepClean(dir)) {
+        warn(s"after deep clean $dir, used space still higher than $freeSpaceThreshold")
+      }
+    }
+    val finishedTime = System.currentTimeMillis()
+    info(s"clean job $dir finished, elapsed time: ${(finishedTime - startTime) / 1000} s.")
+  }
+
+  def main(args: Array[String]): Unit = {
+    do {
+      info(s"start all clean job")
+      val startTime = System.currentTimeMillis()
+      val hasFinished = new CountDownLatch(cacheDirs.length)
+      cacheDirs.foreach { dir =>
+        threadPool.execute(() => {
+          doCleanJob(dir)
+          hasFinished.countDown()
+        })
+      }
+      hasFinished.await()
```

Comment:
Here waiting for all the threads finsh jobs， means if usedtime greater than scheduleInterval, we can not start next cleaner on time.

Reply:
yes, see the warn log. I think it'ok and it's no meaning if schedule interval is too small.

Reply:
I got the idea if two thread cleaner one specific dir may trigger thread problem.LGTM.